### ISSUE TITLE
configure.ac libical detection needs to reject libical that's too old

### ISSUE THREAD

**Issue by elliefm** [cyrusimap/cyrus-imapd#2952]
Just tried building master while my system was still set up for 2.5 (old libical from apt, no cyruslibs).  The libical version was apparently new enough for configure to not report any errors, but then the build fails because we're using new APIs that don't exist in the that version.

This doesn't affect 3.0 or earlier, because they're still using the old deprecated APIs

**Comment by elliefm** [comment]
@ksmurchison what are your thoughts here?

We currently use PKG_CHECK_MODULES to detect libical (without specifying a minimum version).  We then have a bunch of AC_CHECK_* calls on the detected libical to figure out what features it supports, and we turn stuff like VPATCH off if the libical won't support it.

But, these days jmap_ical.c and ical_support.c are using a bunch of API calls that are only in libical 3.0, and just don't compile at all with libical 2.

I guess all the feature detection made sense while libical 3 was in development and there wasn't a version number to test for.  But these days?  My inclination is to change the PKG_CHECK_MODULES call to require libical >= 3.0.0, and discard all the feature-detection stuff.

The alternative would be gating the newer stuff in jmap_ical.c and ical_support.c behind a bunch more feature detection macros, so we can still support libical 2.  But I think I would prefer to have less cruft, not more!

I'm happy to do the slugwork here, just looking for an approach we agree on before sinking time into it :)

**Comment by ksmurchison** [comment]
I guess this depends on how many distros are still shipping pre-3.0 versions.

**Comment by dilyanpalauzov** [comment]
If libical 3.0.5 is going to be required then https://github.com/cyrusimap/cyrus-imapd/pull/2718 can be integtated (trigger an update imip message, when the timezone in DTSTART changes, the relative to the timezone local time does not change, but the effective distance from the UTC midnight does change).

**Comment by elliefm** [comment]
Current debian stable (buster) has 3.0.4.

I don't know how fedora/centos releases/versions work but I did a quick search and it looks like there's been libical 3.0.x since fedora 29 and centos 7.7? (but maybe I just found an archive of rpms built for those versions, idk)

Looking at [libical release notes](https://github.com/libical/libical/blob/master/ReleaseNotes.txt) there's a few new functions in 3.0.x that weren't in 3.0.0, but we're not (currently) using them.  If we wanted to require 3.0.0 we'd still need to feature-detect those if we wanted to use them, but we aren't at the moment (though #2718 would need to)

**Comment by elliefm** [comment]
WIP is here: https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v31/2952-libical-detection

So far I've mostly been down a rabbit hole making the `--disable-jmap` build succeed, but that's dealt with now.  @ksmurchison do you mind reviewing the changes so far, please?

Next, I'll start working on the configure.ac changes.

**Comment by ksmurchison** [comment]
@elliefm The changes look sane to me.  

Do we need to #ifdef protect SORT_SNOOZEDUNTIL as per your changes?  As long as the code where we set it is #ifdef protected, it will never get set without JMAP support, so having it defined or used in switch() shouldn't be a problem, right?

**Comment by elliefm** [comment]
I had to protect the case in `index_msgdata_load()` cause it's calling `jmap_fetch_snoozed()`, which isn't available... and then gcc started warning about it missing a case for an enum value, so I protected the enum value too.

Thinking about it some more, I might be able to just protect the contents of that one case, instead of the whole case, and then the other protections won't be necessary.  Not sure if that's better or not? It'd be a smaller diff with less #ifdef'ery, but it would also leave open the implication that that case isn't special (unless you happen to look at `index_msgdata_load()`), which might be confusing later.
